src/Utilities/Algorithm.hpp

```diff
@@ -0,0 +1,59 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <numeric>
+
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Get the indices in the container `t` of the `K` elements that satisfy
+ * `op` after sorting with respect to `op`.
+ *
+ * Specifically, if `op` is `[](const lhs, const rhs) { return lhs < rhs; }`
+ * then the result will be the indices of the `K` smallest elements in `t`. If
+ * `op` is `[](const lhs, const rhs) { return lhs > rhs; }` then the result will
+ * be the indices of the `K` largest elements in `t`.
+ *
+ * \warning The indices are not sorted by `op` when returned. Use `std::sort` to
+ * sort them as follows:
+ * \code
+ * const auto op = [](const double lhs, const double rhs) { return lhs < rhs; };
+ * std::vector<double> t{-1.0, 7.8, 9.3, -22.8, 9.5};
+ * auto result = indices_of<3>(t, op);
+ * std::sort(result.begin(), result.end(),
+ *           [&t, &op](const size_t lhs, const size_t rhs) {
+ *             return op(t[lhs], t[rhs]);
+ *           });
+ * \endcode
+ */
+template <size_t K, typename T, typename Comparator>
+auto indices_of(const T& t, Comparator op) noexcept -> std::array<size_t, K> {
+  ASSERT(K <= t.size(),
+         "The container whose K elements we're finding must have at least size "
+             << K << " but has size " << t.size());
+  std::array<size_t, K> indices{};
+  std::iota(indices.begin(), indices.end(), 0);
+  const auto get_max_index = [&indices, &t, &op]() {
+    return std::distance(
+        indices.begin(),
+        std::max_element(indices.begin(), indices.end(),
+                         [&t, &op](const size_t lhs, const size_t rhs) {
+                           return op(gsl::at(t, lhs), gsl::at(t, rhs));
+                         }));
+  };
+  auto max_index = static_cast<size_t>(get_max_index());
+  for (size_t i = K; i < t.size(); ++i) {
+    if (op(gsl::at(t, i), gsl::at(t, max_index))) {
```

Comment:
`max_index` indexes into `indices`, not `t`. Should be something like
```
if (op(gsl::at(t, i), gsl::at(t, gsl::at(indices, max_index)))) {
```

This input demonstrates the bug (because 5.2 is compared either against 7.8 or 4.3 depending on the correctness of the indexing):
```
std::vector<double> t{-1.0, 7.8, 9.3, -22.8, 9.5, 4.3, 5.2};
```

Probably this longer input sequence should be used in the tests.